src/dwarf/debug_info.rs

```diff
@@ -221,336 +230,140 @@ pub enum AttrValue<'a> {
-                if 4 <= data.len() {
-                    Some((AttrValue::Unsigned(decode_uword(data) as u64), 4))
-                } else {
-                    None
-                }
-            } else if 8 <= data.len() {
-                Some((AttrValue::Unsigned(decode_udword(data)), 8))
+                Some((AttrValue::Unsigned(data.read_u32()? as u64), 4))
             } else {
-                None
+                Some((AttrValue::Unsigned(data.read_u64()?), 8))
             }
         }
         DW_FORM_block2 => {
-            if 2 <= data.len() {
-                let bytes = decode_uhalf(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
+            let bytes = data.read_u16()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes.into())?), fullsize))
         }
         DW_FORM_block4 => {
-            if 2 <= data.len() {
-                let bytes = decode_uword(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
-        }
-        DW_FORM_data2 => {
-            if 2 <= data.len() {
-                let v = decode_uhalf(data);
-                Some((AttrValue::Unsigned(v as u64), 2))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data4 => {
-            if 4 <= data.len() {
-                let v = decode_uword(data);
-                Some((AttrValue::Unsigned(v as u64), 4))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data8 => {
-            if 8 <= data.len() {
-                let v = decode_udword(data);
-                Some((AttrValue::Unsigned(v), 8))
-            } else {
-                None
-            }
+            let bytes = data.read_u32()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes as usize)?), fullsize))
         }
+        DW_FORM_data2 => Some((AttrValue::Unsigned(data.read_u16()?.into()), 2)),
+        DW_FORM_data4 => Some((AttrValue::Unsigned(data.read_u32()?.into()), 4)),
+        DW_FORM_data8 => Some((AttrValue::Unsigned(data.read_u64()?), 8)),
         DW_FORM_string => {
-            let s = extract_string(data, 0)?;
-            if (s.len() + 1) <= data.len() {
-                Some((AttrValue::String(s), s.len() + 1))
-            } else {
-                None
-            }
+            let string = data.read_cstr()?;
+            let len = string.to_bytes_with_nul().len();
```

Comment:
this seems like a relatively expensive operation, why can't `.read_cstr()` return both string and its total length (including zero byte) to avoid this unnecessary linear operation?

Reply:
ok, I see you are removing it later, so it's a bit moot point

Reply:
Sounds good. I also don't think it actually is expensive or linear: `to_bytes_with_nul` is a cast to a slice
```rust
    pub const fn to_bytes_with_nul(&self) -> &[u8] {
        // SAFETY: Transmuting a slice of `c_char`s to a slice of `u8`s
        // is safe on all supported targets.
        unsafe { &*(&self.inner as *const [c_char] as *const [u8]) }
    }
```

And `slice::len` is nothing more than an offset read from what I gather:
```
    pub const fn len(&self) -> usize {
        ptr::metadata(self)
    }
```